internal/build.go

```diff
@@ -75,3 +82,62 @@ func buildCommandLoaderFilter(bundleMode bool, ignore []string) func(string, os.
 		return util.LoaderFilter{Ignore: ignore}.Apply(abspath, info, depth)
 	}
 }
+
+func assertUniqueRuleIds(rules []util.Rule) error {
+	pathsByVisitedRuleIds := make(map[string]string)
+
+	for _, rule := range rules {
+		if _, ok := pathsByVisitedRuleIds[rule.PublicId]; ok {
+			return fmt.Errorf(
+				"We cannot create a bundle for your custom rules."+
+					"\nThe bundle contains duplicate rules."+
+					"\nDuplicate rules are:"+
+					"\n- %s"+
+					"\n- %s",
+				rule.Path,
+				pathsByVisitedRuleIds[rule.PublicId],
+			)
+		}
+
+		pathsByVisitedRuleIds[rule.PublicId] = rule.Path
+	}
+
+	return nil
+}
+
+func assertRuleIdsWithoutSnykPrefix(rules []util.Rule) error {
+	for _, rule := range rules {
+		if strings.HasPrefix(rule.PublicId, "SNYK-") {
+			return fmt.Errorf(
+				"We cannot create a bundle for your custom rules."+
+					"\nCustom rule cannot have a name that starts with \"SNYK-\": %s",
+				rule.Path,
+			)
+		}
+	}
+
+	return nil
+}
+
+func assertValidRuleIds(paths []string) error {
+	rules, err := util.RetrieveRules(paths)
+	if err != nil {
+		return err
+	}
+
+	for i, rule := range rules {
+		rules[i].PublicId = strings.ToUpper(rule.PublicId)
+	}
+
+	err = assertUniqueRuleIds(rules)
+	if err != nil {
+		return err
+	}
+
+	err = assertRuleIdsWithoutSnykPrefix(rules)
```

Comment:
I'm torn here because O(2*N) = O(N) but if we can easily go through the rules once then I'd personally be tempted to do that instead of go through them once. Is there a reason why you made these into two separate calls? I can see that it's slightly easier to read but is the combined version much harder to read? Just curious what you think, I don't feel too strongly about this

Reply:
From my analysis, both implementations will result in the rule IDs validity assertion being in O(2 * n) = O(n).
- In this PR's implementation, the uniqueness assertion happens first (O(n)), and then separately, the `SNYK-` prefix assertion (O(n)) comes next. O(n) + O(n) = O(2 * n) = O(n)
- In the implementation suggested in your comment, the uniqueness assertion (O(1)) and the `SNYK-` prefix assertion (O(1)) happen both in the same loop iterating over the rule IDs. Since both operations happen for all rule IDs, we get O(2) * n = O(1) * n = O(n).

The number of operations (2 * n) does not change based on the implementation, and the only thing that changes here is whether they all happen in a single loop or 2 loops. From what I can tell, the benefits of merging these 2 assertions into a single loop won't be significant in terms of performance, and as such it'd not be worth the cost of less readable code. If we add additional assertions for rule IDs it would become even more apparent.

Would that be acceptable?

Reply:
> the benefits of merging these 2 assertions into a single loop won't be significant in terms of performance, and as such it'd not be worth the cost of this code being less readable.

I agree with Ofek, I'll always prefer readability over small performance improvement.